server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -506,7 +508,32 @@ protected void doClose() {
         keepAliveReaper.cancel();
     }
 
+    /**
+     * Wraps the listener to avoid sending StackTraces back to the coordinating
+     * node if the `error_trace` header is set to {@code false}. Upon reading we
+     * default to {@code true} to maintain the same behavior as before the change,
+     * due to older nodes not being able to specify whether it needs stack traces.
+     *
+     * @param <T> the type of the response
+     * @param listener the action listener to be wrapped
+     * @return the wrapped action listener
+     */
+    <T> ActionListener<T> maybeWrapListenerForStackTrace(ActionListener<T> listener) {
+        String header = getThreadPool().getThreadContext().getHeaderOrDefault("error_trace", "true");
+        if (header.equals("false")) {
+            return listener.delegateResponse((l, e) -> {
+                ExceptionsHelper.unwrapCausesAndSuppressed(e, err -> {
+                    err.setStackTrace(EMPTY_STACK_TRACE_ARRAY);
+                    return false;
+                });
+                l.onFailure(e);
+            });
+        }
+        return listener;
+    }
+
     public void executeDfsPhase(ShardSearchRequest request, SearchShardTask task, ActionListener<SearchPhaseResult> listener) {
+        listener = maybeWrapListenerForStackTrace(listener);
         final IndexShard shard = getShard(request);
         rewriteAndFetchShardRequest(shard, request, listener.delegateFailure((l, rewritten) -> {
```

Comment:
since we are actually wrapping the listener here anyway, should we just manually wrap it once maybe?
We have a capturing lambda for `delegateFailure` anyway, so maybe just save heap and indirection and go `new ActionListener() {...}` and override both methods by hand?

Reply:
After the changes in https://github.com/elastic/elasticsearch/pull/118266/commits/cf16266dea83b2d396385d041b30bd09ea5e66f0 I'd rather not, we would need to duplicate also the version checks and I am not sure that would result clean.